src/libstd/sys/unix/process/process_common.rs

```diff
@@ -81,6 +81,7 @@ pub struct Command {
     gid: Option<gid_t>,
     saw_nul: bool,
     closures: Vec<Box<dyn FnMut() -> io::Result<()> + Send + Sync>>,
+    groups: Option<Vec<gid_t>>,
```

Comment:
`Option<Vec<_>>` is always a little suspicious; do we need to differentiate between empty and non-existent?

Reply:
I guess I wasn't sure if `setgroups(0, NULL)` is functionally equivalent to `setgroups(0, groups)`.

Reply:
I wrote a quick c program to test it and I think they are the same. So, I think it should be fine to just use `Vec<gid_t>`.
```
#include <grp.h>
#include <unistd.h>
#include <err.h>

int main(void)
{
    gid_t gids[1];
    if (setgroups(0, gids) == -1)
        err(-1, "setgroups");
    if (execl("/usr/bin/id", "id", NULL) == -1)
        err(-1, "execl"); 
    return 0;
}
```

Reply:
We do need to distinguish whether we were given groups to set at all, and if not, we shouldn't call `setgroups` at all. (There's one other case where we do need to, for compatibility; see below for the logic I think we want here.) So, you do need to keep this as an `Option`, but it should be `Option<Box<[gid_t]>>`

Reply:
So there’s a difference between calling `.groups(vec![])` and never calling it to start with?

Reply:
If so, then it seems that we need an "unset groups" method to perform the equivalent of `self.groups = None`.

Reply:
> if `setgroups(0, NULL)` is functionally equivalent to `setgroups(0, groups)`

FWIW, I was arguing for the equivalent of

```rust
if self.groups.is_empty() {
    set_them()
}
```

instead of 

```rust
if let Some(groups) = self.groups {
    set_them()
}
```

Not that you should unconditionally call the C API. Maybe you *need* to unconditionally call it (I don't know this API), but if so, you could do

```rust
let (ptr, count) = if self.groups.is_empty() {
    (nullptr, 0)
} else {
    (groups.as_ptr, groups.len())
};
```

Reply:
To be honest, I'm not too familiar with this API either. I just saw in the previous iteration of the cosw, we were calling `setgroups` unconditionally and ignoring the failure.